### PR TITLE
Cort clear dst layout (fixes for #1204)

### PR DESCRIPTION
**Pull request by cdwfs**
Incorporates feedback from @scyganINTEL's original PR #1204, and improves test coverage a bit.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -8177,6 +8177,69 @@ static bool VerifyDestImageLayout(layer_data *dev_data, GLOBAL_CB_NODE *cb_node,
...
+        for (uint32_t layerIdx = 0; layerIdx < range.layerCount; ++layerIdx) {
+            uint32_t layer = layerIdx + range.baseArrayLayer;
+            VkImageSubresource sub = {range.aspectMask, level, layer};
+            IMAGE_CMD_BUF_LAYOUT_NODE node;
+            if (!FindLayout(cb_node, image, sub, node)) {
+                SetLayout(cb_node, image, sub, IMAGE_CMD_BUF_LAYOUT_NODE(destImageLayout, destImageLayout));
+                continue;
+            }
+            if (node.layout != destImageLayout) {
+                UNIQUE_VALIDATION_ERROR_CODE error_code = VALIDATION_ERROR_00000;
```
Comment (tobine) [inline layers/core_validation.cpp]: Same comment as above here. Default to 01085 & have "else" assert that func_name is "vkCmdClearColorImage()"

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -8177,6 +8177,69 @@ static bool VerifyDestImageLayout(layer_data *dev_data, GLOBAL_CB_NODE *cb_node,
...
+                if (strcmp(func_name, "vkCmdClearColorImage()") == 0) {
+                    error_code = VALIDATION_ERROR_01085;
+                } else if (strcmp(func_name, "vkCmdClearDepthStencilImage()") == 0) {
+                    error_code = VALIDATION_ERROR_01100;
+                } else {
+                    assert(0);
+                }
+                skip_call |= log_msg(
+                    dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_COMMAND_BUFFER_EXT, 0,
+                    __LINE__, error_code, "DS", "Cannot clear an image whose layout is %s and "
```
Comment (tobine) [inline layers/core_validation.cpp]: Also prepend func_name here

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -8177,6 +8177,69 @@ static bool VerifyDestImageLayout(layer_data *dev_data, GLOBAL_CB_NODE *cb_node,
     return skip_call;
 }
 
+static bool VerifyClearImageLayout(layer_data *dev_data, GLOBAL_CB_NODE *cb_node, VkImage image, VkImageSubresourceRange range,
+                                   VkImageLayout destImageLayout, const char *func_name) {
+    bool skip_call = false;
```
Comment (tobine) [inline layers/core_validation.cpp]: Can rename this to "skip" to save chars.

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -8177,6 +8177,69 @@ static bool VerifyDestImageLayout(layer_data *dev_data, GLOBAL_CB_NODE *cb_node,
     return skip_call;
 }
 
+static bool VerifyClearImageLayout(layer_data *dev_data, GLOBAL_CB_NODE *cb_node, VkImage image, VkImageSubresourceRange range,
+                                   VkImageLayout destImageLayout, const char *func_name) {
```
Comment (tobine) [inline layers/core_validation.cpp]: For variable names in internal functions (vs. API intercept functions) prefer no camelCase so destImageLayout->dest_image_layout.

File: tests/layer_validation_tests.cpp [hunk 5]
```diff
@@ -10417,12 +10418,55 @@ TEST_F(VkLayerTest, InvalidImageLayout) {
...
     err = vkCreateImage(m_device->device(), &image_create_info, NULL, &dst_image);
     ASSERT_VK_SUCCESS(err);
+    image_create_info.format = VK_FORMAT_D32_SFLOAT;
+    image_create_info.usage |= VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT;
+    err = vkCreateImage(m_device->device(), &image_create_info, NULL, &depth_image);
+    ASSERT_VK_SUCCESS(err);
+
+    // Allocate memory
+    VkMemoryRequirements img_mem_reqs = {};
+    VkMemoryAllocateInfo memAlloc = {};
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: favor no camelCase for var names

File: tests/layer_validation_tests.cpp [hunk 6]
```diff
@@ -10473,12 +10517,75 @@ TEST_F(VkLayerTest, InvalidImageLayout) {
...
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_01086);
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_01085);
+    m_commandBuffer->ClearColorImage(dst_image, VK_IMAGE_LAYOUT_UNDEFINED, &clearValue, 1, &clearRange);
+    m_errorMonitor->VerifyFound();
+    // Fail due to provided layout not matching actual current layout for color clear.
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_01085);
+    m_commandBuffer->ClearColorImage(dst_image, VK_IMAGE_LAYOUT_GENERAL, &clearValue, 1, &clearRange);
+    m_errorMonitor->VerifyFound();
+    // This one should work correctly
+    m_errorMonitor->ExpectSuccess();
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: In general we try to isolate positive tests and negative tests completely separate.
Based on that I'd say this last check is not needed. I understand that it is useful, though, as we already have the state setup so it's simple to add this final call and verify that it works. Based on that I'm fine keeping it but will defer to @mark-lunarg or @chrisforbes if they strenuously object.

File: tests/layer_validation_tests.cpp [hunk 7]
```diff
@@ -10473,12 +10517,75 @@ TEST_F(VkLayerTest, InvalidImageLayout) {
...
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_01101);
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_01100);
+    m_commandBuffer->ClearDepthStencilImage(depth_image, VK_IMAGE_LAYOUT_UNDEFINED, &depthClearValue, 1, &clearRange);
+    m_errorMonitor->VerifyFound();
+    // Fail due to provided layout not matching actual current layout for depth clear.
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_01100);
+    m_commandBuffer->ClearDepthStencilImage(depth_image, VK_IMAGE_LAYOUT_GENERAL, &depthClearValue, 1, &clearRange);
+    m_errorMonitor->VerifyFound();
+    // This one should work correctly
+    m_errorMonitor->ExpectSuccess();
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Same comment as above. I'm fine with or without this last call.

File: tests/layer_validation_tests.cpp (hunk 6 above)
Comment (mark-lunarg) [inline tests/layer_validation_tests.cpp]: I don't think it's worth all this extra code being moved to info VkPositiveLayerTest, so I agree with Tobine that it isn't needed (here and below).

Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: I think this is OK for now.

**Pull request merged**